crates/ruff_server/src/edit.rs

```diff
@@ -29,6 +31,57 @@ pub enum PositionEncoding {
     UTF8,
 }
 
+/// A unique document ID, derived from a URL passed as part of an LSP request.
+/// This document ID can point to either be a standalone Python file, a full notebook, or a cell within a notebook.
+#[derive(Clone, Debug)]
+pub(crate) enum DocumentKey {
+    Notebook(PathBuf),
+    NotebookCell(lsp_types::Url),
+    Text(PathBuf),
+}
+
+impl DocumentKey {
+    /// Creates a document key from a URL provided in an LSP request.
+    pub(crate) fn from_url(url: &lsp_types::Url) -> Self {
+        if url.scheme() != "file" {
+            return Self::NotebookCell(url.clone());
+        }
+        let Some(path) = url.to_file_path().ok() else {
+            return Self::NotebookCell(url.clone());
+        };
+
+        // figure out whether this is a notebook or a text document
+        if path.extension() == Some(OsStr::new("ipynb")) {
+            Self::Notebook(path)
+        } else {
+            // Until we support additional document types, we need to confirm
+            // that any non-notebook file is a Python file
+            debug_assert_eq!(path.extension(), Some(OsStr::new("py")));
```

Comment:
Based on your comment, does this need to be a runtime assertion or is it fine to keep it in the debug build?

Reply:
This assertion is more of a sanity check for a developer than something we'd actually want to crash the extension over, which is why I've left it in the debug build.